From 3d9ced680581a8481e490e4e790bb87ef1fa8911 Mon Sep 17 00:00:00 2001
From: Wei Qin <qinwei61@huawei.com>
Date: Thu, 16 Jul 2026 16:37:43 +0800
Subject: [PATCH 1/1] udma: bugfix related to dtu_en

driver inclusion
category: bugfix
bugzilla: https://atomgit.com/openeuler/umdk/issues/692
CVE: NA

-----------------------------------------------------------

This patch fix a bug related to dtu_en.

Fixes: a7ea7e683ec6 ("udma: support alloc and free tid")

Signed-off-by: Wei Qin <qinwei61@huawei.com>
---
 src/urma/hw/udma/udma_u_buf.c    |  6 +++---
 src/urma/hw/udma/udma_u_common.h |  4 ++--
 src/urma/hw/udma/udma_u_jetty.c  |  4 ++--
 src/urma/hw/udma/udma_u_jfs.c    |  2 +-
 src/urma/hw/udma/udma_u_jfs.h    |  6 +++---
 src/urma/hw/udma/udma_u_ops.c    | 21 +++++++++++----------
 6 files changed, 22 insertions(+), 21 deletions(-)

diff --git a/src/urma/hw/udma/udma_u_buf.c b/src/urma/hw/udma/udma_u_buf.c
index 2911e08..6dc2a79 100644
--- a/src/urma/hw/udma/udma_u_buf.c
+++ b/src/urma/hw/udma/udma_u_buf.c
@@ -61,7 +61,7 @@ bool udma_u_alloc_queue_buf(struct udma_u_jetty_queue *q, uint32_t max_entry_cnt
 
 	if (q->dtu_en || q->sq_reserved)
 		return true;
-	if (!q->cstm && q->ctx->hugepage_enable) {
+	if (q->ctx->hugepage_enable) {
 		q->hugepage = udma_u_alloc_hugepage(q->ctx, q->qbuf_size);
 		if (q->hugepage) {
 			q->qbuf = q->hugepage->va_start;
@@ -69,7 +69,7 @@ bool udma_u_alloc_queue_buf(struct udma_u_jetty_queue *q, uint32_t max_entry_cnt
 			UDMA_LOG_WARN("hugepage buffer unavailable, fallback to normal buffer.\n");
 			q->qbuf = udma_u_alloc_kernel_buf(q->ctx, q->qbuf_size);
 		}
-	} else if (!q->cstm) {
+	} else {
 		q->qbuf = udma_u_alloc_kernel_buf(q->ctx, q->qbuf_size);
 	}
 
@@ -95,7 +95,7 @@ void udma_u_free_queue_buf(struct udma_u_jetty_queue *q)
 		q->wrid = NULL;
 	}
 
-	if (q->dtu_en || q->cstm || q->sq_reserved)
+	if (q->cstm || q->dtu_en || q->sq_reserved)
 		return;
 
 	if (q->qbuf != NULL) {
diff --git a/src/urma/hw/udma/udma_u_common.h b/src/urma/hw/udma/udma_u_common.h
index 55667da..42f1e28 100644
--- a/src/urma/hw/udma/udma_u_common.h
+++ b/src/urma/hw/udma/udma_u_common.h
@@ -81,8 +81,8 @@ struct udma_u_hugepage {
 	struct udma_u_hugepage_priv *priv;
 };
 
-typedef int (*dtu_u_mmap_func)(uint64_t va_base, uint64_t va_size, void **dtu_va);
-typedef void (*dtu_u_munmap_func)(void **g_dtu_va, uint64_t va_size);
+typedef int (*dtu_u_mmap_func)(uint64_t va_base, uint64_t va_size, void **dtu_va, uint32_t *dtu_va_refcount);
+typedef void (*dtu_u_munmap_func)(void **g_dtu_va, uint64_t va_size, uint32_t *dtu_va_refcount);
 
 /* 32 */
 #define UDMA_JETTY_TABLE_NUM 1 << 5
diff --git a/src/urma/hw/udma/udma_u_jetty.c b/src/urma/hw/udma/udma_u_jetty.c
index e0bc48b..a3b8142 100644
--- a/src/urma/hw/udma/udma_u_jetty.c
+++ b/src/urma/hw/udma/udma_u_jetty.c
@@ -1141,8 +1141,8 @@ urma_status_t udma_u_active_jetty(urma_jetty_t *jetty)
 	urma_cmd_udrv_priv_t udata = {};
 	int ret;
 
-    udma_jetty->sq.is_wqe_lock_buf_jetty =
-		is_support_wqe_lock_buffer_jetty(udma_jetty->sq.ctx, jetty->jetty_cfg.id, true);
+    udma_jetty->sq.is_wqe_lock_buf_jetty = 
+		is_support_wqe_lock_buffer_jetty(udma_jetty->sq.ctx, jetty->jetty_cfg.id);
 	if (udma_jetty->sq.is_wqe_lock_buf_jetty) {
 		if (udma_u_check_wqe_lock_buffer_jetty(udma_jetty))
 			return URMA_EINVAL;
diff --git a/src/urma/hw/udma/udma_u_jfs.c b/src/urma/hw/udma/udma_u_jfs.c
index 1d024df..00fffff 100644
--- a/src/urma/hw/udma/udma_u_jfs.c
+++ b/src/urma/hw/udma/udma_u_jfs.c
@@ -991,7 +991,7 @@ urma_status_t udma_u_post_one_wr(struct udma_u_context *udma_ctx,
 
 	if (udma_check_sge_num_and_opcode(wr->opcode, sq, wr, &wqe_info.opcode,
 					  udma_ctx->atomic_add_en)) {
-		UDMA_LOG_ERR("WR SGE number or opcode is invalid.\n");
+		UDMA_LOG_ERR("WR SGE number or opcode: %u is invalid.\n", wr->opcode);
 		return URMA_EINVAL;
 	}
 
diff --git a/src/urma/hw/udma/udma_u_jfs.h b/src/urma/hw/udma/udma_u_jfs.h
index f29164c..803d77c 100644
--- a/src/urma/hw/udma/udma_u_jfs.h
+++ b/src/urma/hw/udma/udma_u_jfs.h
@@ -177,15 +177,15 @@ static inline uint32_t get_max_sge_num(uint8_t max_sge, uint32_t max_inline_size
 	return UDMA_MAX(max_sge, size);
 }
 
-static inline bool is_support_wqe_lock_buffer_jetty(struct udma_u_context *udma_ctx,
-						    uint32_t id, bool is_jetty)
+static inline bool is_support_wqe_lock_buffer_jetty(struct udma_u_context *udma_ctx, 
+						    uint32_t id)
 {
 	bool is_ccu_jetty;
 
 	is_ccu_jetty = udma_ctx->ccu_jetty_max_cnt &&
 		       id >= udma_ctx->ccu_jetty_start_id &&
 		       id < (udma_ctx->ccu_jetty_start_id + udma_ctx->ccu_jetty_max_cnt);
-	return is_jetty && is_ccu_jetty && udma_ctx->lock_buffer_en;
+	return is_ccu_jetty && udma_ctx->lock_buffer_en;
 }
 
 urma_status_t udma_u_post_sq_wr(struct udma_u_context *udma_ctx,
diff --git a/src/urma/hw/udma/udma_u_ops.c b/src/urma/hw/udma/udma_u_ops.c
index 969e6f8..0e7a431 100644
--- a/src/urma/hw/udma/udma_u_ops.c
+++ b/src/urma/hw/udma/udma_u_ops.c
@@ -27,6 +27,7 @@
 
 #define LIBDTU_FILE "/usr/lib64/libdtu.so"
 static void *g_dtu_va = NULL;
+static uint32_t g_dtu_va_refcount = 0;
 static pthread_mutex_t g_sq_reserved_lock = PTHREAD_MUTEX_INITIALIZER;
 static uint64_t g_sq_reserved_refcount = 0;
 static void *g_sq_reserved_va = NULL;
@@ -255,12 +256,8 @@ err_mmap_dlsym:
 
 static void udma_close_dtu_provider(struct udma_u_context *udma_ctx)
 {
-	if (udma_ctx->dtu_munmap_fun_ptr) {
-		udma_ctx->dtu_munmap_fun_ptr(&g_dtu_va, udma_ctx->dtu_va_size);
-		udma_ctx->dtu_munmap_fun_ptr = NULL;
-	}
+	udma_ctx->dtu_munmap_fun_ptr = NULL;
 	udma_ctx->dtu_mmap_fun_ptr = NULL;
-	udma_ctx->dtu_enable = false;
 	dlclose(udma_ctx->dtu_provider_handle);
 }
 
@@ -304,9 +301,10 @@ static urma_context_t *udma_u_create_context(urma_device_t *dev, uint32_t eid_in
 	udma_u_init_context(udma_ctx, &resp);
 	if (udma_ctx->dtu_enable) {
 		if (udma_open_dtu_provider(udma_ctx)) {
-			if (udma_ctx->dtu_mmap_fun_ptr(udma_ctx->dtu_va_base, udma_ctx->dtu_va_size, &g_dtu_va)) {
-				udma_ctx->dtu_mmap_fun_ptr = NULL;
+			if (udma_ctx->dtu_mmap_fun_ptr(udma_ctx->dtu_va_base,
+			    udma_ctx->dtu_va_size, &g_dtu_va, &g_dtu_va_refcount)) {
 				udma_close_dtu_provider(udma_ctx);
+				udma_ctx->dtu_enable = false;
 				UDMA_LOG_WARN("Could not mmap dtu, continuing with the normal process.\n");
 			}
 		} else {
@@ -330,8 +328,10 @@ static urma_context_t *udma_u_create_context(urma_device_t *dev, uint32_t eid_in
 err_reserved_sq:
 	udma_u_free_db(&udma_ctx->urma_ctx, &udma_ctx->db);
 err_alloc_db:
-	if (udma_ctx->dtu_enable)
+	if (udma_ctx->dtu_enable) {
+		udma_ctx->dtu_munmap_fun_ptr(&g_dtu_va, udma_ctx->dtu_va_size, &g_dtu_va_refcount);
 		udma_close_dtu_provider(udma_ctx);
+	}
 	(void)urma_cmd_delete_context(&udma_ctx->urma_ctx);
 err_create_ctx:
 	(void)pthread_mutex_destroy(&udma_ctx->hugepage_lock);
@@ -351,9 +351,10 @@ static urma_status_t udma_u_delete_context(urma_context_t *ctx)
 	udma_u_destroy_jt_table(udma_ctx);
 	udma_u_free_reserved_sq();
 	udma_u_free_db(ctx, &udma_ctx->db);
-	if (udma_ctx->dtu_enable)
+	if (udma_ctx->dtu_enable) {
+		udma_ctx->dtu_munmap_fun_ptr(&g_dtu_va, udma_ctx->dtu_va_size, &g_dtu_va_refcount);
 		udma_close_dtu_provider(udma_ctx);
-
+	}
 	udma_u_destroy_hugepage(udma_ctx);
 
 	if (urma_cmd_delete_context(&udma_ctx->urma_ctx)) {
-- 
2.28.0.windows.1

